### ISSUE TITLE
Import FormsModule in CoreModule or SharedModule?

### ISSUE THREAD

**Issue by vikct** [zaarheed/angular5-starter-multi-module#1]
Hi sir, first of all thank you for your guide. I've been following this app structure since i begun my angular training. I do understand that CoreModule should behave like AppModule, but instead it wont get bootstrapped at the start. I've been trying to import FormsModule into CoreModule, to be used in 1 of my public module, instead it returned error. but if i imported it in SharedModule, it works. to my understanding, SharedModule is called when there's shared services, directives, components created under it. it seems like both SharedModule has overtaken some of the CoreModule role, mind if you provide some clarifications on it? thanks

**Comment by oputyk** [comment]
Hi sir.

Firstly let me explain something. If module A wants service X, then module
A looks for the service in it's providers array. If there is no service X
provided in the module A, then it looks for the service in the module's A
parent. And if the module's A parent doesn't have the service X provided
too then the module's A parent's parent is asked about the service X and so
on. It is like a hierarchy. After all if there is no service X provided
then there is thrown an error.

And now, the SharedModule is for shared components, pipes but not services.
If you need to put a shared service it means you want to have one instance
per a module (the same for every component in this module and that's why
I'm this case you put the service in the module's "provided" array) instead
of creating for each component new one instance of it (but if you expect
this behavior than it is good to put the service somewhere, maybe in the
SharedModule and provide it for some modules or components independently).

The CoreModule is for providing all shared services (singletons and it
works because if other modules imported in the CoreModule want to use a
service then they will look for it in their parent modules and in this case
they will find it in the CoreModule). Also CoreModule's responsibility is
to import the other forms modules.

I hope you understand now how it works and what it should look like.

Best regards, Kamil Cegliński.

**Comment by vikct** [comment]
Still dont get it.

I've tried import FormsModule in AppModule, which is the root of the application, yet it still returning error when i attempted to call ngModel.

So in this case, SharedModule behave like the central libraries for all shared modules, including libraries from node modules and custom modules/directives/services?

**Comment by oputyk** [comment]
Firstly I don't see error log and whole code you have. Secondly I am
beginner too :)
So if you have some troubles create a post on Stackoverflow and send me
link to it. Maybe I will be able to help you or others will.

**Comment by zaarheed** [comment]
Hey @vikct,

Keep in mind this structure is experimental, and it's how I think things should be done for the applications I build - but not necessarily suitable for everyone else. You should follow it as a guide to get you started, but make decisions suited for your application at every step along the way :)

I would import `FormsModule` in to your `PublicModule` or whatever the equivalent is. I would not import it in to `AppModule` because I believe this needs to be kept as clean as possible. I would not import it in to `CoreModule` unless one of the components in `CoreModule` has a dependency on `FormsModule`.

I prefer the approach of importing `FormsModule` in to `PublicModule` (or equivalent) because that it where your dependency for `FormsModule` is. If one day you decide to remove the `PublicModule` feature completely, you do not want lingering dependencies if you can help it. The idea is to encapsulate each module as much as possible so that it can operate (and be tested) completely independently.

Try importing the `FormsModule` in to your feature module and let me know how you get on.

`SharedModule` should contain pipes, components (and some services) that are shared across your modules. You can import `FormsModule` in to the `SharedModule` if more than one module in your component has a dependency on `FormsModule`. However, there is a trade-off so you need to decide if you should import `FormsModule` in to each feature module separately, or import it into `SharedModule` and have it available to all your feature modules.

@oputyk makes a very good point about putting services in the `SharedModule`. It is suitable for services which are not required as singletons in your application (such as services which define HTTP endpoints or retrieve data). However if a service is storing any type of state that needs to persist across the entirety of your application, it should not be in the `SharedModule` as each time it is imported in to a feature module it is a separate instance.

**TL;DR:** To answer @vikct's question more directly: yes you can import 'FormsModule' in to 'SharedModule' because it doesn't need to be a singleton in your application. However, I would only import 'FormsModule' in to 'SharedModule' if it is used across multiple modules in your application. If 'FormsModule' is a dependency on only one feature module, then import it in to that `FeatureModule` directly.

I hope I have addressed your concern!

**Comment by vikct** [comment]
good explanation!
first off, this is a great structure imo
The thing is that, i'm still new to angular, i need every guidance i can get on why i can do this, and why i can't. 
I've tried importing the FormsModule into the AppModule, but it's not a matter of wouldnt, it will return error still.
How about importing Angular Material? in your opinion, how would you handle these imports? Again, I imported them in SharedModule.

**Comment by zaarheed** [comment]
You shouldn't import it in to `AppModule` because nothing runs through that module, it's purely used to bootstrap the application and keep in line with the format the Angular CLI likes.

I would import Angular Material in to each module which uses it. Naturally, I expect you are using Angular Material everywhere in your application so yes the `SharedModule` is the correct place for this.

**Comment by vikct** [comment]
thank you so much sir! this clears much doubt!
I will continue to use your module as a basis for my angular development, i really agree with the way you structured the whole thing!
